### PR TITLE
Separate ir tree generation from semantic checking

### PR DESCRIPTION
**Pull request by jdconrad**
This change creates a new phase to generate an ir tree once semantic checking is complete. The phase depends on the decorations generated during semantic checking against the user tree. Separation of the phases in this fashion will allow us to not generate an ir tree in the future when it's not necessary for features such as auto-completion or initial stored script compilation.

While the change is long, the majority follows the pattern of removing the ir tree generation from each user tree node and instead adding it the UserTreeToIRTreeVisitor along with some new decorations to support this.

### CODE REVIEW

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/ir/ClassNode.java [hunk 1]
```diff
@@ -24,8 +24,8 @@
 import org.elasticsearch.painless.MethodWriter;
 import org.elasticsearch.painless.ScriptClassInfo;
 import org.elasticsearch.painless.phase.IRTreeVisitor;
-import org.elasticsearch.painless.symbol.WriteScope;
```
Comment (stu-elastic) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/ir/ClassNode.java:27]: How did this pass style checking?

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/ir/InstanceofNode.java [hunk 2]
```diff
@@ -22,50 +22,28 @@
...
+import org.elasticsearch.painless.lookup.def;
 import org.elasticsearch.painless.phase.IRTreeVisitor;
 import org.elasticsearch.painless.symbol.WriteScope;
-import org.objectweb.asm.Type;
 
 public class InstanceofNode extends UnaryNode {
     
     /* ---- begin node data ---- */
-
+
```
Comment (stu-elastic) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/ir/InstanceofNode.java:32]: Fix this ws issue.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/node/AExpression.java [hunk 3]
```diff
@@ -91,20 +79,12 @@ public PainlessCast cast(SemanticScope semanticScope) {
...
 
-        return AnalyzerCaster.getLegalCast(getLocation(), valueType, targetType, isExplicitCast, isInternalCast);
-    }
+        PainlessCast painlessCast = AnalyzerCaster.getLegalCast(getLocation(), valueType, targetType, isExplicitCast, isInternalCast);
 
-    static ExpressionNode cast(ExpressionNode expressionNode, PainlessCast painlessCast) {
-        if (painlessCast == null) {
-            return expressionNode;
+        if (painlessCast != null) {
+            semanticScope.putDecoration(this, new ExpressionPainlessCast(painlessCast));
```
Comment (stu-elastic) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/node/AExpression.java:85]: Note: Node now decorated with cast.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/node/EBool.java [hunk 4]
```diff
@@ -21,9 +21,6 @@
 
 import org.elasticsearch.painless.Location;
 import org.elasticsearch.painless.Operation;
-import org.elasticsearch.painless.ir.BooleanNode;
-import org.elasticsearch.painless.ir.ClassNode;
-import org.elasticsearch.painless.lookup.PainlessCast;
 import org.elasticsearch.painless.phase.UserTreeVisitor;
 import org.elasticsearch.painless.symbol.Decorations.Read;
 import org.elasticsearch.painless.symbol.Decorations.TargetType;
```
Comment (stu-elastic) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/node/EBool.java:26]: Rename to `EBoolComparison` or similar

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/node/EBoolean.java [hunk 5]
```diff
@@ -20,10 +20,9 @@
 package org.elasticsearch.painless.node;
```
Comment (stu-elastic) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/node/EBoolean.java:20]: Rename to `EBoolConst` on similar.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/node/ERegex.java [hunk 6]
```diff
@@ -105,87 +91,7 @@ Output analyze(ClassNode classNode, SemanticScope semanticScope) {
...
-
-        MemberFieldLoadNode memberFieldLoadNode = new MemberFieldLoadNode();
-        memberFieldLoadNode.setLocation(getLocation());
-        memberFieldLoadNode.setExpressionType(Pattern.class);
-        memberFieldLoadNode.setName(name);
-        memberFieldLoadNode.setStatic(true);
-        output.expressionNode = memberFieldLoadNode;
-
-        return output;
+        semanticScope.putDecoration(this, new StandardConstant(flags));
```
Comment (stu-elastic) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/node/ERegex.java:94]: Note: Moved to IR generation.

**Review by stu-elastic — APPROVED** [review]
Great change.

**Comment by jdconrad** [comment]
@stu-elastic Thanks for the review. Will respond to PR comments soon.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/ir/ClassNode.java (hunk 1 above)
Comment (jdconrad) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/ir/ClassNode.java:27]: I don't think we check order. I "optimized" imports for all the files in Intellij so hopefully it picks this up.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/node/EBool.java (hunk 4 above)
Comment (jdconrad) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/node/EBool.java:26]: Changed.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/node/EBoolean.java (hunk 5 above)
Comment (jdconrad) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/node/EBoolean.java:20]: Changed.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/node/ERegex.java (hunk 6 above)
Comment (jdconrad) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/node/ERegex.java:94]: Note for @stu-elastic

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/ir/InstanceofNode.java (hunk 2 above)
Comment (jdconrad) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/ir/InstanceofNode.java:32]: Fixed for consistency with all the other nodes based on existing blanks lines and sections delineated by comments.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/node/AExpression.java (hunk 3 above)
Comment (jdconrad) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/node/AExpression.java:85]: Note for @stu-elastic

**Pull request merged**